Added documentation for the stack bulk actions

dnaploszek opened the pull request:
# Description of the change

Added documentation for the upcoming bulk actions redesign, with stack list bulk actions first. 

## Checklist

Please make sure that the proposed change checks all the boxes below before requesting a review:

- [x] I have reviewed the [guidelines for contributing](https://github.com/spacelift-io/user-documentation/blob/main/CONTRIBUTING.md) to this repository.
- [x] The preview looks fine.
- [x] The tests pass.
- [x] The commit history is clean and meaningful.
- [x] The pull request is opened against the `main` branch.
- [ ] The pull request is no longer marked as a draft.
- [ ] You agree to license your contribution under the [MIT license](https://github.com/spacelift-io/user-documentation/blob/main/LICENSE) to Spacelift (not required for Spacelift employees).
- You have updated the navigation files correctly:
    - [ ] No new pages have been added, or;
    - [ ] Only _nav.yaml_ has been updated because the changes only apply to SaaS, or;
    - [ ] Only _nav.self-hosted.yaml_ has been updated because the changes only apply to Self-Hosted, or;
    - [x] Both _nav.yaml_ and _nav.self-hosted.yaml_ have been updated.

If the proposed change is ready to be merged, please request a review from `@spacelift-io/solutions-engineering`. Someone will review and merge the pull request.

_Spacelift employees should request reviews from the relevant engineers and are allowed to merge pull requests after they got at least one approval._

Thank you for your contribution! 🙇

dnaploszek commented on docs/concepts/stack/stack-bulk-actions.md:
```diff
@@ -0,0 +1,44 @@
+!!! info
+    The UI will only show the actions that can be performed on the selected stacks filtering out the unavailable ones for your convenience.
+
+## How to use bulk actions
```
I'm not sure if that's the right place for this description, as this is going to be common across all bulk actions instances. So would love to get an idea of where I could put it 🤔

adamconnelly replied:
I think we've got a few options here:

1. We could have a single page for bulk actions instead of embedding it in the stacks section. It could then have different sections for the different types of actions that can be performed.
2. We could have a bulk actions page that explains the concept in general terms, and then in the stacks page you would just explain the stack-specific actions.
3. Or we just accept the duplication on the grounds that it's simpler for users to have the specific info for the resource type they're viewing in one place. The downside to this for us is more places to maintain if we change bulk actions.

Maybe @jmfontaine has some thoughts about this?

dnaploszek replied:
I went with the 2nd option

adamconnelly reviewed (commented):
I'll post some more specific wording suggestions separately, but I just wanted to post some general thoughts first.

General notes:

- The screenshots of the stack list are quite blurry when I view them on the preview. I wonder if there's a way to improve that, or if we should maybe use something like this [lightbox plugin](https://squidfunk.github.io/mkdocs-material/reference/images/#lightbox).
- Let's split this page up into a few sections. That way we can give the information to people in more bite-sized chunks, we can use the table of contents to highlight the functionality to users, and it's handy to be able to directly link to info for support.

How about something like the following structure:

```
# Stack bulk actions
<intro goes here>

## Available actions
<list the actions along with what they can do>

## How to use bulk actions
### Performing actions

The basic flow with the floating dialog would go here.

### Action details

Info about how to use the sidebar to get more details.

### Stopping actions

How to stop in-progress actions.
```

adamconnelly commented on docs/concepts/stack/stack-bulk-actions.md:
```diff
@@ -0,0 +1,45 @@
+# Stack bulk actions
+
+Spacelift allows you to perform bulk actions on stacks. This is useful when you need to perform the same action on multiple stacks at once. To initiate bulk actions, navigate to the "Stacks" page and select the stacks you want to perform the action on. When you make the selection you can choose the intended action from the following options:
+
```
A couple of general thoughts about the action list:

- Maybe a table would look better than a list for this information?
- Should we sort the items in alphabetical order? The reason I say that is the list is relatively long, so it would make it easier to find specific items.

dnaploszek replied:
Is there any table that you could point me towards that you pictured? I'm assuming you want two rows, one with the action name and the other with the description? Looking at the other files I've not seen a lot of tables so I just went with the list 🤔.

dnaploszek replied:
As for the order, I wanted to keep the similar order as the actions would appear on the bar (from most popular to least), but alphabetical is fine too :)

adamconnelly replied:
Sure - we've got one in the self-hosted section for the provider compatibility matrix: https://docs.spacelift.io/self-hosted/latest/vendors/terraform/terraform-provider. I think there might be some elsewhere, but I'm not sure off the top of my head.

Yeah - I was thinking two columns, one for the action name and another for the description, with a row per action.

But please just treat this as a suggestion - if you don't think it's easier to read the information with the table that's fine. I just thought with so many items it might make it easier to scan.

adamconnelly replied:
RE the order - that makes perfect sense, and both options are valid. I just know if I'm reading docs like this and searching for something (vs in a UI when I'm actually using it, and it makes sense for the most used options to be sorted first) it's usually easier for me to find if it's alphabetical.

dnaploszek replied:
Yeah after the consideration I think I'll pass on the table, for the other bulk actions there's going to be much less items, so not sure if the table will make it clearer. Plus after your suggestions it's a bit more concise and easier to scan through 👌🏻

adamconnelly replied:
No problem - sounds good.

adamconnelly commented on docs/concepts/stack/stack-bulk-actions.md:
```diff
@@ -0,0 +1,45 @@
+
+Spacelift allows you to perform bulk actions on stacks. This is useful when you need to perform the same action on multiple stacks at once. To initiate bulk actions, navigate to the "Stacks" page and select the stacks you want to perform the action on. When you make the selection you can choose the intended action from the following options:
+
+- Confirm - Allows to confirm the blocking run for the selected stacks that are in the `Unconfirmed` state.
```
```suggestion
- Confirm - Confirms the blocking run for stacks in the `Unconfirmed` state.
```

adamconnelly commented on docs/concepts/stack/stack-bulk-actions.md:
```diff
@@ -0,0 +1,45 @@
+Spacelift allows you to perform bulk actions on stacks. This is useful when you need to perform the same action on multiple stacks at once. To initiate bulk actions, navigate to the "Stacks" page and select the stacks you want to perform the action on. When you make the selection you can choose the intended action from the following options:
+
+- Confirm - Allows to confirm the blocking run for the selected stacks that are in the `Unconfirmed` state.
+- Discard - Allows to discard the blocking run for the selected stacks that are in the `Unconfirmed` state.
```
```suggestion
- Discard - Discards the blocking run for stacks in the `Unconfirmed` state.
```

adamconnelly commented on docs/concepts/stack/stack-bulk-actions.md:
```diff
@@ -0,0 +1,45 @@
+
+- Confirm - Allows to confirm the blocking run for the selected stacks that are in the `Unconfirmed` state.
+- Discard - Allows to discard the blocking run for the selected stacks that are in the `Unconfirmed` state.
+- Trigger - Allows to [manually trigger a tracked run](../run/tracked.md#triggering-manually) for the selected stacks.
```
```suggestion
- Trigger - [Triggers a tracked run](../run/tracked.md#triggering-manually) for the selected stacks.
```

adamconnelly commented on docs/concepts/stack/stack-bulk-actions.md:
```diff
@@ -0,0 +1,45 @@
+- Confirm - Allows to confirm the blocking run for the selected stacks that are in the `Unconfirmed` state.
+- Discard - Allows to discard the blocking run for the selected stacks that are in the `Unconfirmed` state.
+- Trigger - Allows to [manually trigger a tracked run](../run/tracked.md#triggering-manually) for the selected stacks.
+- Sync commit - Allows to sync the tracked stack commit. It pulls the latest HEAD commit of the tracked branch and updates the tracked commit field.
```
```suggestion
- Sync commit - Updates the selected stacks to point at the latest HEAD commit for their tracked branches.
```

adamconnelly commented on docs/concepts/stack/stack-bulk-actions.md:
```diff
@@ -0,0 +1,45 @@
+- Discard - Allows to discard the blocking run for the selected stacks that are in the `Unconfirmed` state.
+- Trigger - Allows to [manually trigger a tracked run](../run/tracked.md#triggering-manually) for the selected stacks.
+- Sync commit - Allows to sync the tracked stack commit. It pulls the latest HEAD commit of the tracked branch and updates the tracked commit field.
+- Lock - Allows to lock the selected stacks for exclusive use.
```
```suggestion
- Lock - Locks the selected stacks for exclusive use.
```

adamconnelly commented on docs/concepts/stack/stack-bulk-actions.md:
```diff
@@ -0,0 +1,45 @@
+- Trigger - Allows to [manually trigger a tracked run](../run/tracked.md#triggering-manually) for the selected stacks.
+- Sync commit - Allows to sync the tracked stack commit. It pulls the latest HEAD commit of the tracked branch and updates the tracked commit field.
+- Lock - Allows to lock the selected stacks for exclusive use.
+- Unlock - Allows to unlock the selected locked stacks.
```
```suggestion
- Unlock - Unlocks the selected stacks that are currently locked.
```

adamconnelly commented on docs/concepts/stack/stack-bulk-actions.md:
```diff
@@ -0,0 +1,45 @@
+- Sync commit - Allows to sync the tracked stack commit. It pulls the latest HEAD commit of the tracked branch and updates the tracked commit field.
+- Lock - Allows to lock the selected stacks for exclusive use.
+- Unlock - Allows to unlock the selected locked stacks.
+- Enable - Allows to enable the selected disabled stacks.
```
```suggestion
- Enable - Enables the selected stacks.
```

adamconnelly commented on docs/concepts/stack/stack-bulk-actions.md:
```diff
@@ -0,0 +1,45 @@
+- Lock - Allows to lock the selected stacks for exclusive use.
+- Unlock - Allows to unlock the selected locked stacks.
+- Enable - Allows to enable the selected disabled stacks.
+- Disable - Allows to disable the selected stacks, so they will not trigger any runs.
```
```suggestion
- Disable - Disables the selected stacks, so they will not trigger any runs.
```

adamconnelly commented on docs/concepts/stack/stack-bulk-actions.md:
```diff
@@ -0,0 +1,45 @@
+- Unlock - Allows to unlock the selected locked stacks.
+- Enable - Allows to enable the selected disabled stacks.
+- Disable - Allows to disable the selected stacks, so they will not trigger any runs.
+- Approve - Allows to add a run review approval for the selected stacks that require reviews according to your [approval policy](../policy/approval-policy.md).
```
```suggestion
- Approve - Adds an approval for the selected stacks that require [approval policy](../policy/approval-policy.md) reviews.
```

adamconnelly commented on docs/concepts/stack/stack-bulk-actions.md:
```diff
@@ -0,0 +1,45 @@
+- Enable - Allows to enable the selected disabled stacks.
+- Disable - Allows to disable the selected stacks, so they will not trigger any runs.
+- Approve - Allows to add a run review approval for the selected stacks that require reviews according to your [approval policy](../policy/approval-policy.md).
+- Reject - Allows to add a run review approval for the selected stacks that require reviews according to your [approval policy](../policy/approval-policy.md).
```
```suggestion
- Reject - Adds a rejection for the selected stacks that require [approval policy](../policy/approval-policy.md) reviews.
```

adamconnelly commented on docs/concepts/stack/stack-bulk-actions.md:
```diff
@@ -0,0 +1,45 @@
+- Disable - Allows to disable the selected stacks, so they will not trigger any runs.
+- Approve - Allows to add a run review approval for the selected stacks that require reviews according to your [approval policy](../policy/approval-policy.md).
+- Reject - Allows to add a run review approval for the selected stacks that require reviews according to your [approval policy](../policy/approval-policy.md).
+- Run task - Allows to [manually trigger a run](../run/task.md) with a custom command.
```
```suggestion
- Run task - [Manually triggers a run](../run/task.md) with a custom command.
```

adamconnelly commented on docs/concepts/stack/stack-bulk-actions.md:
```diff
@@ -0,0 +1,45 @@
+- Run task - Allows to [manually trigger a run](../run/task.md) with a custom command.
+
+!!! info
+    The UI will only show the actions that can be performed on the selected stacks filtering out the unavailable ones for your convenience.
```
```suggestion
    The UI will only show the actions that can be performed on the selected stacks, filtering out the unavailable ones for your convenience.
```

adamconnelly commented on docs/concepts/stack/stack-bulk-actions.md:
```diff
@@ -0,0 +1,45 @@
+
+## How to use bulk actions
+
+When you select stacks on the "Stacks" page a floating action bar will appear at the bottom of the screen. This bar will show the number of selected stacks and the available actions.
```
```suggestion
After you select one or more stacks, a floating action bar will appear at the bottom of the screen. This bar will show the number of selected stacks and the available actions.
```

adamconnelly commented on docs/concepts/stack/stack-bulk-actions.md:
```diff
@@ -0,0 +1,45 @@
+On the drawer you will see a fine-grained details on which actions are available for each of the selected stacks.
+![](../../assets/screenshots/stack/bulk-actions/stack-bulk-actions_partial-action-drawer.png)
+
+Once you select action, you will be presented with the confirmation step, that allows you to add additional details (like a note for the Lock action) and confirm the action itself.
```
```suggestion
Once you select an action, you will be presented with a confirmation step that allows you to add additional details (like a note for the Lock action) and to confirm the action itself.
```

adamconnelly commented on docs/concepts/stack/stack-bulk-actions.md:
```diff
@@ -0,0 +1,45 @@
+The same view is also available on the drawer, where you can review the applicable and not applicable items again.
+![](../../assets/screenshots/stack/bulk-actions/stack-bulk-actions_confirm-action-drawer.png)
+
+Once you confirm the action you'll be presented with the action results drawer, where you can review the status of each item. Please stay on this view until all actions are done otherwise you will stop the execution of the actions on all the selected items. When the action execution is finished, you can use the "New action" button to perform another action on the same selection or a subset of it  ...
```
```suggestion
Once you confirm the action you'll be presented with the action results drawer, where you can review the status of each item. Please stay on this view until all of your actions are performed, otherwise any in-progress actions will be stopped.

When the action execution is finished, you can use the "New action" button to perform another action on the same selection or a subset of it (from the completed or the failed results).
```

adamconnelly commented on nav.self-hosted.yaml:
```diff
@@ -7,6 +7,7 @@ nav:
           - concepts/stack/README.md
           - concepts/stack/creating-a-stack.md
           - concepts/stack/stack-settings.md
+          - concepts/stack/stack-bulk-actions.md
```
Just want to check @dnaploszek - is this functionality currently feature flagged? If so can you hold off adding it to the Self-Hosted navigation until the flag is removed, and the self-hosted frontend submodule is updated to pull in the latest changes.

Otherwise it'll end up in the docs, but the functionality won't actually be available for Self-Hosted customers.

dnaploszek replied:
Yes it's feature flagged so I'll remove it for now and add it to the [SH preparation task](https://app.clickup.com/t/8695axy38). Thanks!

adamconnelly replied:
:bow:

adamconnelly reviewed (commented):
All the content is good - I've just left a load of suggestions to simplify some of the wording.

adamconnelly commented on docs/concepts/stack/stack-bulk-actions.md:
```diff
@@ -0,0 +1,60 @@
+# Stack bulk actions
+
+Spacelift allows you to perform bulk actions on stacks. This is useful when you need to perform the same action on multiple stacks at once. To initiate bulk actions, navigate to the "Stacks" page and select the stacks you want to perform the action on. When you make the selection you can choose the intended action from the following options:
```
Maybe move the "When you make the selection you can choose..." sentence into the "Available actions" section. It's just slightly weird to have a `:` and then a heading following it.

adamconnelly commented on docs/concepts/stack/stack-bulk-actions.md:
```diff
@@ -0,0 +1,60 @@
+
+### Performing actions
+
+If you need a more detailed view of the selected stacks, click the "See details" button, that will open the bulk actions drawer. From the drawer you can also dismiss any stacks you deem unnecessary. Use floating bar for quick actions and use the drawer when you need to be more careful with your selection.
```
```suggestion
If you need a more detailed view of the selected stacks, click the "See details" button, that will open the bulk actions drawer. From the drawer you can also dismiss any stacks you deem unnecessary. Use the floating bar for quick actions and use the drawer when you need to be more careful with your selection.
```

adamconnelly commented on docs/concepts/stack/stack-bulk-actions.md:
```diff
@@ -0,0 +1,60 @@
+
+![](../../assets/screenshots/stack/bulk-actions/stack-bulk-actions_drawer.png)
+
+On the drawer you will see a fine-grained details on which actions are available for each of the selected stacks.
```
```suggestion
On the drawer you will see fine-grained details on which actions are available for each of the selected stacks.
```

adamconnelly commented on docs/concepts/stack/stack-bulk-actions.md:
```diff
@@ -0,0 +1,60 @@
+
+Once you confirm the action you'll be presented with the action results drawer, where you can review the status of each item. Please stay on this view until all of your actions are performed, otherwise any in-progress actions will be stopped.
+
+When the action execution is finished, you can use the "New action" button to perform another action on the same selection or a subset of it (from the completed or the failed results).
```
```suggestion
When execution is complete, you can use the "New action" button to perform another action on the same selection or a subset of it (from the completed or the failed results).
```

I just felt like the word "action" was repeated too many times here. I think my suggestion is a little clumsy, so if you can think of a better option then feel free.

dnaploszek replied:
`when the bulk execution is complete` 🤷🏻 ?

adamconnelly replied:
Yeah - that sounds good. :+1:

adamconnelly reviewed (commented):
Just spotted a few really minor things while re-reviewing.

dnaploszek commented:
@adamconnelly I've updated the screenshots to be more focused so it reduced the blurriness. I've also went for a slight structure change where I moved bulk actions into a separate page under "Product" as we're going to extend that page with the other upcoming bulk actions.

adamconnelly commented:
> @adamconnelly I've updated the screenshots to be more focused so it reduced the blurriness. I've also went for a slight structure change where I moved bulk actions into a separate page under "Product" as we're going to extend that page with the other upcoming bulk actions.

Sounds good - I'll re-review later today.

adamconnelly reviewed (approved):
Looks good. Thanks for making the changes. The screenshots look much better now as well.
